java/monkey/src/test/java/ParserTest.java

```diff
@@ -77,14 +68,200 @@ void testString() {
+                        "((a + b) - c)"
+                ),
+                List.of(
+                        "a * b * c",
+                        "((a * b) * c)"
+                ),
+                List.of(
+                        "a * b / c",
+                        "((a * b) / c)"
+                ),
+                List.of(
+                        "a + b / c",
+                        "(a + (b / c))"
+                ),
+                List.of(
+                        "a + b * c + d / e - f",
+                        "(((a + (b * c)) + (d / e)) - f)"
+                ),
+                List.of(
+                        "3 + 4; -5 * 5",
+                        "(3 + 4)\n((-5) * 5)" // TODO Here, we are inserting a \n since there are two Statements. Should we stick to the book?
+                ),
+                List.of(
+                        "5 > 4 == 3 < 4",
+                        "((5 > 4) == (3 < 4))"
+                ),
+                List.of(
+                        "5 < 4 != 3 > 4",
+                        "((5 < 4) != (3 > 4))"
+                ),
+                List.of(
+                        "3 + 4 * 5 == 3 * 1 + 4 * 5",
+                        "((3 + (4 * 5)) == ((3 * 1) + (4 * 5)))"
+                ),
+                List.of(
+                        "3 + 4 * 5 == 3 * 1 + 4 * 5",
+                        "((3 + (4 * 5)) == ((3 * 1) + (4 * 5)))"
+                )
+        );
+
+        for (List<String> test : tests) {
+            var program = buildProgram(test.get(0));
+
+            Assertions.assertEquals(test.get(1), program.toString());
+        }
+    }
+
+    private void testIntegerLiteral(Expression expression, Long expectedValue) {
+        if (expression instanceof IntegerLiteralExpression integerLiteralExpression) {
+            Assertions.assertEquals(expectedValue, integerLiteralExpression.getValue());
+            Assertions.assertEquals(expectedValue.toString(), integerLiteralExpression.tokenLiteral());
+        } else {
+            throw new AssertionError(expression.getClass().getSimpleName() + " is not instance of IntegerLiteralExpression");
+        }
+    }
+
+    private Program buildProgram(String input) {
+        var l = new Lexer(input);
+        var p = new Parser(l);
+        var program = p.parseProgram();
+
+        checkParseErrors(p);
+
+        return program;
+    }
+
+    private void testLetStatement(Statement statement, String name) {
         Assertions.assertEquals(TokenType.LET.token().literal(), statement.tokenLiteral());
 
         if (statement instanceof LetStatement letStatement) {
             Assertions.assertEquals(name, letStatement.getName().getValue());
             Assertions.assertEquals(name, letStatement.getName().tokenLiteral());
         } else {
-            throw new AssertionError("statement not instance of LetStatement");
+            throw new AssertionError(statement.getClass().getSimpleName() + " is not instance of LetStatement");
```

Comment:
There might be an Assertions for Classes? I know there is one for Exceptions.

Reply:
There is one, I just wanted to cast and verify at the same time. I used the class asserting in some tests to keep nesting to a minimal.